samcli/local/docker/container.py

```diff
@@ -314,6 +334,32 @@ def wait_for_logs(self, stdout=None, stderr=None):
 
         self._write_container_output(logs_itr, stdout=stdout, stderr=stderr)
 
+    def wait_for_socket_connection(self):
+        """
+        Waits for a successful connection to the socket used to communicate with Docker.
+        """
+        start_time = time.time()
+        sleep = 0.1
+        while True:
+            a_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            location = (self._container_host, self.rapid_port_host)
+            # connect_ex returns 0 if connection succeeded
+            is_port_open = not a_socket.connect_ex(location)
+            a_socket.close()
+
+            if is_port_open:
+                break
+
+            current_time = time.time()
+            if current_time - start_time > START_CONTAINER_TIMEOUT:
+                raise ContainerStartTimeoutException(
+                    f"Timed out while starting container. You can increase this timeout by "
+                    f"setting the SAM_CLI_START_CONTAINER_TIMEOUT environment variable. "
+                    f"The current timeout is {START_CONTAINER_TIMEOUT} (seconds)."
```

Comment:
Not sure that this message makes sense anymore, as this wait is not part of the start container flow anymore. This is still related to starting a container, since after the container is started, it can take a bit of time for the socket to start listening. But this is also an independent/standalone function for waiting for a socket connection, so I wonder if I should update this to `Timed out while attempting to establish a connection to the container ...` or something like that (and the corresponding variable/constant/etc changes). In the end, we want a balance between an accurate message but also one that is clear for users. Thoughts?

Reply:
This method is being used by `wait_for_result()` only, which is intended to be used by different use cases. The exception `ContainerStartTimeoutException` can be misleading here. Should we rename the exception be `ContainerTimeoutException` or equivalent and using the message that you suggested?

> But this is also an independent/standalone function for waiting for a socket connection

Is it the intent of this method? From the use case in this change, it seems to be a private method of `Container` class to wait for a connection to be ready to represent a container is ready to serve traffic. Is this method designed to be used outside of `Container` class at all? If no then may be adding an underscore prefix is better?

Reply:
Agreed to change the messaging (`Timed out while attempting to establish a connection to the container ...`) and also the exception name (`ContainerTimeoutException`) to align with that.

Reply:
should `START_CONTAINER_TIMEOUT` also be changed to another name?

Reply:
done 👍